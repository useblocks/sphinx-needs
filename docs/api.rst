.. _api:

Python API
==========

**Sphinx-Needs** provides an open API for other Sphinx-extensions to provide specific need-types, create needs or
make usage of the filter possibilities.

The API allows the injection of extra configuration, but
does not support manipulation of it (e.g remove need types),
to keep the final configuration transparent for the Sphinx project authors.

.. _api_configuration:

Configuration
-------------

.. automodule:: sphinx_needs.api.configuration
   :members:

.. autoclass:: sphinx_needs.functions.functions.DynamicFunction
   :members: __name__,__call__

Need
----

.. automodule:: sphinx_needs.api.need
   :members:


Exceptions
----------

.. automodule:: sphinx_needs.api.exceptions
   :members:

Data
----

.. automodule:: sphinx_needs.data
   :members: NeedsInfoType, NeedsMutable, NeedsPartType

Views
-----

These views are returned by certain functions, and injected into filters,
but should not be instantiated directly.

.. automodule:: sphinx_needs.views
   :members:
   :undoc-members:
   :special-members: __iter__, __getitem__, __len__
