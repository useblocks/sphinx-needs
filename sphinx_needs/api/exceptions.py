from sphinx.errors import SphinxError, SphinxWarning


class NeedsApiConfigException(SphinxError):
    """
    A configuration changes collides with the already provided configuration by the user.

    Example: An extension wants to add an already existing needs_type.
    """


class NeedsApiConfigWarning(SphinxWarning):
    pass


class NeedsNoIdException(SphinxError):
    pass


class NeedsDuplicatedId(SphinxError):
    pass


class NeedsStatusNotAllowed(SphinxError):
    pass


class NeedsTagNotAllowed(SphinxError):
    pass


class NeedsInvalidException(SphinxError):
    pass


class NeedsConstraintNotAllowed(SphinxError):
    pass


class NeedsConstraintFailed(SphinxError):
    pass


class NeedsInvalidOption(SphinxError):
    pass


class NeedsTemplateException(SphinxError):
    pass


class NeedsInvalidFilter(SphinxError):
    pass
