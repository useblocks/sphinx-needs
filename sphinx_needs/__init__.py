from sphinx_needs.needs import __version__, setup  # noqa: F401
