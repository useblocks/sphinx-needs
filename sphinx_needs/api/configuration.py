"""
API to get or add specific sphinx needs configuration parameters.

All functions here are available under ``sphinxcontrib.api``. So do not import this module directly.
"""
from typing import Any, Callable, List, Optional

from docutils.parsers.rst import directives
from sphinx.application import Sphinx
from sphinx.util.logging import SphinxLoggerAdapter

from sphinx_needs.api.exceptions import NeedsApiConfigException, NeedsApiConfigWarning
from sphinx_needs.config import NEEDS_CONFIG
from sphinx_needs.functions import register_func
from sphinx_needs.functions.functions import DynamicFunction


def get_need_types(app: Sphinx) -> List[str]:
    """
    Returns a list of directive-names from all configured need_types.

    **Usage**::

        from sphinx_needs.api import get_need_types

        all_types = get_need_types(app)

    :param app: Sphinx application object
    :return: list of strings
    """
    needs_types = app.config.needs_types
    return [x["directive"] for x in needs_types]


def add_need_type(
    app: Sphinx, directive: str, title: str, prefix: str, color: str = "#ffffff", style: str = "node"
) -> None:
    """
    Adds a new need_type to the configuration.

    The given directive must no exist, otherwise NeedsApiConfigException gets raised.

    Same impact as using :ref:`needs_types` manually.

    **Usage**::

        from sphinx_needs.api import add_need_type

        add_need_type(app, 'awesome', 'Awesome', 'AW_', '#000000', 'cloud')

    :param app: Sphinx application object
    :param directive: Name of the directive, e.g. 'story'
    :param title: Long, human-readable title, e.g. 'User-Story'
    :param prefix: Prefix, if IDs get automatically generated. E.g.: ``'US_'``
    :param color: Hex-color code used in needflow representation. Default: ``'#ffffff'``
    :param style: Plantuml-style for needflow representation. Default: 'node'
    :return: None
    """
    import sphinx_needs.directives.need

    needs_types = app.config.needs_types
    type_names = [x["directive"] for x in needs_types]

    if directive in type_names:
        raise NeedsApiConfigException(f"{directive} already exists as need type")

    needs_types.append({"directive": directive, "title": title, "prefix": prefix, "color": color, "style": style})
    app.add_directive(directive, sphinx_needs.directives.need.NeedDirective)


def add_extra_option(app: Sphinx, name: str) -> None:
    """
    Adds an extra option to the configuration. This option can then later be used inside needs or ``add_need``.

    Same impact as using :ref:`needs_extra_options` manually.

    **Usage**::

        from sphinx_needs.api import add_extra_option

        add_extra_option(app, 'my_extra_option')

    :param app: Sphinx application object
    :param name: Name as string of the extra option
    :return: None
    """

    extra_options = NEEDS_CONFIG.create_or_get("extra_options", dict)

    if name in extra_options:
        raise NeedsApiConfigWarning(f"Option {name} already registered.")

    NEEDS_CONFIG.add("extra_options", {name: directives.unchanged}, dict, append=True)


def add_dynamic_function(app: Sphinx, function: DynamicFunction, name: Optional[str] = None) -> None:
    """
    Registers a new dynamic function for sphinx-needs.

    If ``name`` is not given, the name to call the function is automatically taken from the provided function.
    The used name must be unique.

    **Usage**::

        from sphinx_needs.api import add_dynamic_function

        def my_function(app, need, needs, *args, **kwargs):
            # Do magic here
            return "some data"

        add_dynamic_function(app, my_function)

    Read :ref:`dynamic_functions` for details about how to use dynamic functions.

    :param app: Sphinx application object
    :param function: Function to register
    :param name: Name of the dynamic function as string
    :return: None
    """
    register_func(function, name)


# 'Need' is untyped, so we temporarily use 'Any' here
WarningCheck = Callable[[Any, SphinxLoggerAdapter], bool]


def add_warning(
    app: Sphinx, name: str, function: Optional[WarningCheck] = None, filter_string: Optional[str] = None
) -> None:
    """
    Registers a warning.

    A warning can be based on the result of a given filter_string or an own defined function.

    :param app: Sphinx app object
    :param name: Name as string for the warning
    :param function: function to execute to check the warning
    :param filter_string: filter_string to use for the warning
    :return: None
    """
    warnings_option = NEEDS_CONFIG.create_or_get("warnings", dict)

    if function is None and filter_string is None:
        raise NeedsApiConfigException("Function or filter_string must be given for add_warning_func")

    if function is not None and filter_string is not None:
        raise NeedsApiConfigException(
            "For add_warning_func only function or filter_string is allowed to be set, " "not both."
        )

    warning_check = function or filter_string

    if name in warnings_option:
        raise NeedsApiConfigException(f"Warning {name} already registered.")

    NEEDS_CONFIG.add("warnings", {name: warning_check}, dict, append=True)
