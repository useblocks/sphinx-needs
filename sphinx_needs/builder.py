import os
from typing import Iterable, Optional, Set

from docutils import nodes
from sphinx import version_info
from sphinx.application import Sphinx
from sphinx.builders import Builder

from sphinx_needs.logging import get_logger
from sphinx_needs.needsfile import NeedsList
from sphinx_needs.utils import unwrap

log = get_logger(__name__)


class NeedsBuilder(Builder):
    name = "needs"
    format = "json"
    file_suffix = ".txt"
    links_suffix = None

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needs = env.needs_all_needs.values()  # We need a list of needs for later filter checks
        filters = env.needs_all_filters
        config = env.config
        version = getattr(config, "version", "unset")
        needs_list = NeedsList(config, self.outdir, self.srcdir)

        if config.needs_file:
            needs_file = config.needs_file
            needs_list.load_json(needs_file)
        else:
            # check if needs.json file exists in conf.py directory
            needs_json = os.path.join(self.srcdir, "needs.json")
            if os.path.exists(needs_json):
                log.info("needs.json found, but will not be used because needs_file not configured.")

        # Clean needs_list from already stored needs of the current version.
        # This is needed as needs could have been removed from documentation and if this is the case,
        # removed needs would stay in needs_list, if list gets not cleaned.
        needs_list.wipe_version(version)
        #
        from sphinx_needs.filter_common import filter_needs

        filter_string = self.app.config.needs_builder_filter
        filtered_needs = filter_needs(self.app, needs, filter_string)

        for need in filtered_needs:
            needs_list.add_need(version, need)

        for need_filter in filters.values():
            if need_filter["export_id"]:
                needs_list.add_filter(version, need_filter)

        try:
            needs_list.write_json()
        except Exception as e:
            log.error(f"Error during writing json file: {e}")
        else:
            log.info("Needs successfully exported")

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needs_json(app: Sphinx, _exception: Exception) -> None:

    env = unwrap(app.env)

    if not env.config.needs_build_json:
        return

    # Do not create an additional needs.json, if builder is already "needs".
    if isinstance(app.builder, NeedsBuilder):
        return

    try:
        needs_builder = NeedsBuilder(app, env)
    except TypeError:
        needs_builder = NeedsBuilder(app)
        needs_builder.set_environment(env)

    needs_builder.finish()


class NeedumlsBuilder(Builder):
    name = "needumls"

    def write_doc(self, docname: str, doctree: nodes.document) -> None:
        pass

    def finish(self) -> None:
        env = unwrap(self.env)
        needumls = env.needs_all_needumls.values()

        for needuml in needumls:
            if needuml["save"]:
                puml_content = needuml["content_calculated"]
                # check if given save path dir exists
                save_path = os.path.join(self.outdir, needuml["save"])
                save_dir = os.path.dirname(save_path)
                if not os.path.exists(save_dir):
                    os.makedirs(save_dir, exist_ok=True)

                log.info(f"Storing needuml data to file {save_path}.")
                with open(save_path, "w") as f:
                    f.write(puml_content)

    def get_outdated_docs(self) -> Iterable[str]:
        return []

    def prepare_writing(self, _docnames: Set[str]) -> None:
        pass

    def write_doc_serialized(self, _docname: str, _doctree: nodes.document) -> None:
        pass

    def cleanup(self) -> None:
        pass

    def get_target_uri(self, _docname: str, _typ: Optional[str] = None) -> str:
        return ""


def build_needumls_pumls(app: Sphinx, _exception: Exception) -> None:
    env = unwrap(app.env)

    if not env.config.needs_build_needumls:
        return

    # Do not create additional files for saved plantuml content, if builder is already "needumls".
    if isinstance(app.builder, NeedumlsBuilder):
        return

    # if other builder like html used together with config: needs_build_needumls
    if version_info[0] >= 5:
        needs_builder = NeedumlsBuilder(app, env)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
    else:
        needs_builder = NeedumlsBuilder(app)
        needs_builder.outdir = os.path.join(needs_builder.outdir, env.config.needs_build_needumls)
        needs_builder.set_environment(env)

    needs_builder.finish()
