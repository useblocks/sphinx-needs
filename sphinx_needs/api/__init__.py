# flake8: noqa

from .configuration import (
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id
