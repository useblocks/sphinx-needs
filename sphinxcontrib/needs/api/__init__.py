from .configuration import (  # noqa: F401
    add_dynamic_function,
    add_extra_option,
    add_need_type,
    get_need_types,
)
from .need import add_external_need, add_need, del_need, make_hashed_id  # noqa: F401
