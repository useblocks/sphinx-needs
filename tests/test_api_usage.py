import sys
import types
from pathlib import Path

import pytest
from sphinx.testing.util import SphinxTestApp

from sphinx_needs.api import add_need_type, get_need_types


@pytest.fixture()
def add_extension1():
    dummy_code = """
def setup(app):
    from sphinx_needs.api import get_need_types

    def after_config(app, config):
        print(get_need_types(app))

    # app.connect('config-inited', after_config)
    print(get_need_types(app))
    return {'version': '0.1'}
    """
    dummy_extension = types.ModuleType("dummy_extension.dummy")
    exec(dummy_code, dummy_extension.__dict__)
    sys.modules["dummy_extension.dummy"] = dummy_extension
    yield
    sys.modules.pop("dummy_extension.dummy")


@pytest.fixture()
def add_extension2():
    dummy_code = """
def setup(app):
    return {'version': '0.1'}
"""
    dummy_extension = types.ModuleType("dummy_extension.dummy")
    exec(dummy_code, dummy_extension.__dict__)
    sys.modules["dummy_extension.dummy"] = dummy_extension
    yield
    sys.modules.pop("dummy_extension.dummy")


@pytest.mark.usefixtures("add_extension1")
@pytest.mark.parametrize(
    "test_app",
    [
        {
            "buildername": "html",
            "srcdir": "doc_test/doc_basic",
            "no_plantuml": True,
            "confoverrides": {"extensions": ["sphinx_needs", "dummy_extension.dummy"]},
        }
    ],
    indirect=True,
)
def test_api_configuration(test_app: SphinxTestApp):
    app = test_app

    app.build()
    assert app._warning.getvalue() == ""

    html = Path(app.outdir, "index.html").read_text()
    assert html is not None


@pytest.mark.usefixtures("add_extension2")
@pytest.mark.parametrize(
    "test_app",
    [
        {
            "buildername": "html",
            "srcdir": "doc_test/doc_basic",
            "no_plantuml": True,
            "confoverrides": {"extensions": ["sphinx_needs", "dummy_extension.dummy"]},
        }
    ],
    indirect=True,
)
def test_api_get_types(test_app: SphinxTestApp):
    app = test_app

    need_types = get_need_types(app)
    assert set(need_types) == {"story", "spec", "impl", "test"}


@pytest.mark.usefixtures("add_extension2")
@pytest.mark.parametrize(
    "test_app",
    [
        {
            "buildername": "html",
            "srcdir": "doc_test/doc_basic",
            "no_plantuml": True,
            "confoverrides": {"extensions": ["sphinx_needs", "dummy_extension.dummy"]},
        }
    ],
    indirect=True,
)
def test_api_add_type(test_app: SphinxTestApp, snapshot):
    app = test_app

    add_need_type(app, "awesome", "Awesome", "AW_", "#000000", "cloud")
    need_types = app.config.needs_types
    assert need_types == snapshot

    Path(app.srcdir, "other_api.rst").write_text(
        """
:orphan:

.. awesome:: my awesome need
    :id: AW_001
        """
    )

    app.build()
    assert app._warning.getvalue() == ""

    html = Path(app.outdir, "other_api.html").read_text()
    assert html is not None
    assert "my awesome need" in html
