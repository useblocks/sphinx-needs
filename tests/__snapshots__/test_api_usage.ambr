# serializer version: 1
# name: test_api_add_type[test_app0]
  list([
    dict({
      'color': '#BFD8D2',
      'directive': 'story',
      'prefix': 'US_',
      'title': 'User Story',
    }),
    dict({
      'color': '#FEDCD2',
      'directive': 'spec',
      'prefix': 'SP_',
      'title': 'Specification',
    }),
    dict({
      'color': '#DF744A',
      'directive': 'impl',
      'prefix': 'IM_',
      'title': 'Implementation',
    }),
    dict({
      'color': '#DCB239',
      'directive': 'test',
      'prefix': 'TC_',
      'title': 'Test Case',
    }),
    dict({
      'color': '#000000',
      'directive': 'awesome',
      'prefix': 'AW_',
      'style': 'cloud',
      'title': 'Awesome',
    }),
  ])
# ---
